### PR TITLE
Geospatial map with case list filters

### PR DESCRIPTION
**Pull request by SmittieC**
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-2780)

## Product Description

Added Case List Filters to the Geospatial map. Nothing is shown yet.

## Technical Summary

Added standard case filters for the geospatial map by subclassing `ProjectReport` and `CaseListMixin` (a bunch of boillerplate code like the `CaseManagementDispatcher` etc is there to make this possible). This filters will ultimately determine which cases shows up on the map. We need to determine which case property to consider the location of the case.

The reason I didn't just subclass [GenericMapReport](https://github.com/dimagi/commcare-hq/blob/92656e0bb2f324a9aa9e9a79aacb6fb0b3a8aa52/corehq/apps/reports/standard/maps.py#L14) is
1. I want the display layer to be handled by JS/HTML alone (easier for me at least)
2. The code didn't really make sense to me.
3. I wanted to keep things as simple as possible for now to keep things flexibile for future work (coming soon)
4. We can refactor later if we see that we can easily subclass from `GenericMapReport`

The way that report data is being passed to the JS is...interesting, but I followed [this example](https://github.com/dimagi/commcare-hq/blob/5ca2ffe38023b0412c906d1aac1830bf0e4fa032/corehq/apps/reports/static/reports/js/base_maps.js). Totally open to better ways if this is considered not-good.

## Feature Flag

Geospatial

### Safety story

- Local testing
- Will test on Staging as well

### Automated test coverage

No test coverage, since
1. I'm building on top of existing code following existing patterns
2. There's no business logic yet

### QA Plan

No QA plan for this. Once we show stuff on the map, then we can start the QA process.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### Screenshot
We get a 403 to mapbox. I strongly suspect this is the reason why the map is not showing. We might be restricting IP addresses on mapbox? (If nothing shows on staging, I'll start to worry)

image

### CODE REVIEW

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            return self;
+        };
+
+        var map;
+
+        $(document).ajaxComplete(function () {
+            var mapDiv = $('#geospatial-map');
+            var $data = $(".map-data");
+
+            if (mapDiv.length && !map) {
```
Comment (SmittieC) [inline corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:107]: Adding the `!map` check, stuff works fine when loading the page intially. When applying the filters, the map disappears. Not sure why? When removing this check, I can see it tries to make multiple calls to mapbox. cc @orangejenny do you perhaps have wisdom for me here?

File: corehq/apps/geospatial/models.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
-from shapely.geometry import Point
+from dataclasses import dataclass
```
Comment (SmittieC) [inline corehq/apps/geospatial/models.py:1]: Apologies, I may have snuck in a run of `isort`. I forgot to do that in a separate commit

**Comment by kaapstorm** [comment]
> We need to determine which case property to consider the location of the case.

Feedback from @sravfeyn: For now the case property for GPS coordinates is "commcare_gps_point", and in future we can manage [GPS case properties using the Data Dictionary](https://confluence.dimagi.com/display/USH/Storing+GPS+Case+Properties+in+Elasticsearch+as+GeoPoints).

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            map.on('click', (event) => {
+                let coordinates = getCoordinates(event);
+                moveMarkerToClickedCoordinate(coordinates);
+            });
+            return self;
+        };
+
+        var map;
+
+        $(document).ajaxComplete(function () {
```
Comment (sravfeyn) [inline corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:103]: A comment here will be useful of what this means (it means the report data is loaded?)

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                }
+                clickedMarker = new mapboxgl.Marker({color: "FF0000", draggable: true});
+                clickedMarker.setLngLat(coordinates);
+                clickedMarker.addTo(map);
+            }
+
+            function showCreateCasePopup(coordinates) {
+                new mapboxgl.Popup()
+                .setLngLat(coordinates)
+                .setHTML(coordinates + '<br/><button> Create Case at this point </button>')
```
Comment (sravfeyn) [inline corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:89]: Would it be possible to do this using KnockoutJS?

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 1 above)
Comment (sravfeyn) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:107]: Here is my understanding.

I think the reason that map is disappearing starting from second report load is that the map container is inside the reportcontent block which gets removed when the Apply button is clicked (I haven't verified this hypothesis). Subsequently the map doesn't loaded, because `map` variable already was defined in the first load, so not meeting `!map` condition.

For a solution, I am not 100% sure that below would work, but we could try one of the below

1. Unset the map variable completely whenever the user clicks 'Apply' button, you can do this by subscribing to click event of the 'Apply' button `$(document).on('click', '#apply-filters'...`. This should result in `!map` condition getting met on every report load.
2. Add the mapbox container outside of reportcontent block, in which case it shouldn't get cleared on subsequent loads. But in this case, you would have to reset the map by removing any markers on it that were previously there and any other data.

2 might be better if it works, since that way we don't have to re-render the Mapbox each time. To be consistent with the user experience of removing the table, we can hide the map on click event and show it back on ajaxComplete event with just the data updated.

Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:118]: @sravfeyn Thanks for the suggestions.

> I think the reason that map is disappearing starting from second report load is that the map container is inside the reportcontent block which gets removed when the Apply button is clicked 

I couldn't find evidence for this, but it does seem to be the behaviour. I couldn't get 2 to work, and doing 1 (remounting) seemed to fix it, but in a non-ideal way: Multiple requests to mapbox are being made and sometimes the map still disappears, but not very often. I think this has to do with the order in which events fire. I updated the code for reference, but it will need to be "fixed". Any ideas?

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 3 above)
Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:114]: [255282d](https://github.com/dimagi/commcare-hq/pull/33020/commits/255282dae4f3c1602f1bc76cc39a211dae8bf54e)

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 4 above)
Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:92]: I'm not sure how we would do that here?

Comment (sravfeyn) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:92]: I think I picked a wrong line, but generally I was suggesting to use KnockoutJS observables where we need to propagate or update changes to make the code manageable. 

But given that the PR is already quite challenging we can consider that in a latter PR.

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 1 above)
Comment (sravfeyn) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:119]: To close the loop here, we offlined and got 2 to work.

Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:125]: [We added](https://github.com/dimagi/commcare-hq/pull/33020/commits/03a7a6692320d0de26d3f98cb0a7b188ee146985) a geospatial specific html base. I also updated the code to read coordinates from the `commcare_gps_point` property (format is "lat lng alt acc").

Here's a screenshot (1 case has a coordinate):
image

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:106]: don't see a screen shot including this button, but my guess is it should use `btn btn-default` classes

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            };
+
+            function updateArea(e) {
+                const data = draw.getAll();
+                const area = turf.area(data);
+                // Restrict the area to 2 decimal points.
+                const rounded_area = Math.round(area * 100) / 100;
+            }
+
+            // We should consider refactoring and splitting the below out to a new JS file
```
Comment (orangejenny) [inline corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:55]: Agreed - it doesn't need to block this PR, but it could be quite useful to have relatively independent mapping util functions that could be used here but also in other maps in HQ (thinking specifically of web apps).

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 4 above)
Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js]: Apologies, this stuff is actually not used at the moment. I included this code for when we need to implement creating cases (it's coming), but I'll remove it rather. Removed: [5fab144](https://github.com/dimagi/commcare-hq/pull/33020/commits/5fab1448824f511b5a817e35414cfdbe2cd00c9a)

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 5 above)
Comment (SmittieC) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js:55]: Yeah. We can see how to move it into the existing `map_utils.js` file

File: corehq/apps/geospatial/static/geospatial/js/geospatial_map.js (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/geospatial/static/geospatial/js/geospatial_map.js]: cool, thanks

**Comment by Charl1996** [comment]
@SmittieC It doesn't look like this is on staging yet?

**Comment by SmittieC** [comment]
> @SmittieC It doesn't look like this is on staging yet?

Yeah not on staging. Doesn't make sense to put in on there yet, since we can't do anything except look at the maps. Since it's behind a FF, its safe not not test this on staging yet

File: corehq/apps/geospatial/dispatchers.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    cls_to_view_login_and_domain,
+)
+
+
+class CaseManagementMapDispatcher(ReportDispatcher):
+    prefix = 'geospatial'
+    map_name = 'GEOSPATIAL_MAP'
+
+    @cls_to_view_login_and_domain
+    def dispatch(self, request, *args, **kwargs):
```
Comment (sravfeyn) [inline corehq/apps/geospatial/dispatchers.py:13]: We will need to add a feature-flag check here.

File: corehq/apps/geospatial/dispatchers.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        return super(CaseManagementMapDispatcher, self).dispatch(request, *args, **kwargs)
+
+    def permissions_check(self, report, request, domain=None, is_navigation_check=False):
+        return toggles.GEOSPATIAL.enabled(domain)
```
Comment (sravfeyn) [inline corehq/apps/geospatial/dispatchers.py:17]: This can be removed after https://github.com/dimagi/commcare-hq/pull/33020/files#r1212829624.

File: corehq/apps/geospatial/dispatchers.py (hunk 6 above)
Comment (SmittieC) [reply corehq/apps/geospatial/dispatchers.py:15]: Updated :+1:

File: corehq/apps/geospatial/dispatchers.py (hunk 7 above)
Comment (SmittieC) [reply corehq/apps/geospatial/dispatchers.py]: Updated :+1:

**Review by sravfeyn — APPROVED** [review]

File: corehq/apps/geospatial/reports.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        return context
+
+    def _get_geo_location(self, case):
+        geo_point = case['case_json'].get('commcare_gps_point')
+        if not geo_point:
+            return
+        try:
+            # Update if we need altitude and accuracy
+            lat, lon, _alt, _acc = geo_point.split(" ")
+            return {"lat": float(lat), "lng": float(lon)}
```
Comment (sravfeyn) [inline corehq/apps/geospatial/reports.py:36]: It looks this ends up at `marker.setLngLat` which accepts dict with keys `lon` and `lat` per [the docs](https://docs.mapbox.com/mapbox-gl-js/api/geography/#lnglatlike-example). How come this still works?

Comment (SmittieC) [reply corehq/apps/geospatial/reports.py:36]: In that exaple they used the lon/lat pair, but the description of the LngLatLike object says that it will accept either lon/lat and lng/lat pairs

Comment (sravfeyn) [reply corehq/apps/geospatial/reports.py:36]: Oh yes, thanks!
